core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -108,8 +108,59 @@ class TransactionStateManager(brokerId: Int,
     }
   }
 
-  def enablePidExpiration() {
-    // TODO: add producer id expiration logic
+  def enableTransactionalIdExpiration() {
+    scheduler.schedule("transactionalId-expiration", () => {
+      inLock(stateLock) {
+        val now = time.milliseconds()
+
+        val transactionalIdByPartition: Map[Int, mutable.Iterable[String]] = transactionMetadataCache.flatMap { case (_, entry) =>
+          entry.metadataPerTransactionalId.filter { case (_, txnMetadata) =>
+            txnMetadata.txnLastUpdateTimestamp <= now - config.transactionalIdExpirationMs
+          }.map { case (transactionalId, _) =>
+            transactionalId
+          }
+        }.groupBy { transactionalId =>
+          partitionFor(transactionalId)
+        }
+
+        val recordsPerPartition = transactionalIdByPartition
+          .map { case (partition, transactionalIds) =>
+            val deletes: Array[SimpleRecord] = transactionalIds.map { transactionalId =>
+              new SimpleRecord(now, TransactionLog.keyToBytes(transactionalId), null)
+            }.toArray
+            val records = MemoryRecords.withRecords(TransactionLog.EnforcedCompressionType, deletes: _*)
+            val topicPartition = new TopicPartition(Topic.TRANSACTION_STATE_TOPIC_NAME, partition)
+            (topicPartition, records)
+          }
+
+        def removeFromCacheCallback(responses: collection.Map[TopicPartition, PartitionResponse]): Unit ={
+          responses.foreach { case (topicPartition, response) =>
+            response.error match {
+              case Errors.NONE =>
+                inLock(stateLock) {
+                  val idsToRemove = transactionalIdByPartition(topicPartition.partition())
+                  transactionMetadataCache.get(topicPartition.partition)
+                  .foreach{ txnMetadataEntry =>
+                    idsToRemove.foreach { transactionalId =>
+                      txnMetadataEntry.metadataPerTransactionalId.remove(transactionalId)}
+                  }
+                }
+              case _ =>
+                debug(s"writing transactionalId tombstones for partition: ${topicPartition.partition} failed with error: ${response.error.message()}")
+            }
+          }
+        }
+
+        replicaManager.appendRecords(
```

Comment:
Actually I was wrong before: we do need to hold the state lock to make sure the retrieval of the metadata object from the cache and the append call is in the same monitoring block: https://github.com/apache/kafka/pull/3082/files#r117879684

Note that we need to make sure that when we grab locks, we should always grab the statelock for the whole metadata map cache, and then the lock on the metadata object itself. So in the `removeFromCacheCallback` we cannot grab the lock on statelock since we are already grabbing the lock on the metadata object itself in `delayedOperation.tryComplete()`, and this out-of-ordering will cause dead-lock.

Instead, we can just take the `readlock` after #3082 so that this out-or-ordering will be Okay, note that `metadataPerTransactionalId` is a `Pool` which is thread-safe so `readlock` is sufficient.